python/tank/bootstrap/resolver.py

```diff
@@ -255,31 +264,48 @@ def resolve_shotgun_configuration(
             # resolve primary and user config
             primary_config = None
             user_config = None
+            secondary_primary_config = None
+            secondary_user_config = None
+
             for pc in pipeline_configs:
 
                 # make sure configuration matches our entry point
 
-                if self.__match_plugin_id(pc.get("plugin_ids")) or self.__match_plugin_id(pc.get("sg_plugin_ids")):
+                if self._match_plugin_id(pc.get("plugin_ids")) or self._match_plugin_id(pc.get("sg_plugin_ids")):
                     # we have a matching pipeline configuration!
 
-                    if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
-                        if primary_config:
-                            log.warning(
-                                "More than one pipeline config detected. Will use the most "
-                                "recently updated one."
-                            )
-                        primary_config = pc
+                    if pc["project"] == self._proj_entity_dict:
+                        # this is a direct match
+                        if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
+                            log.debug("Primary match: %s" % pc)
+                            primary_config = pc
+                        else:
+                            user_config = pc
+                            log.debug("Per-user match: %s" % pc)
+
                     else:
-                        # user config
-                        if user_config:
-                            log.warning(
-                                "More than one user config detected. Will use the most "
-                                "recently updated one."
-                            )
-                        user_config = pc
+                        # this is the site level fallback
+                        if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
+                            secondary_primary_config = pc
+                            log.debug("Site level override match: %s" % pc)
+                        else:
+                            secondary_user_config = pc
+                            log.debug("Site level per-user match: %s" % pc)
+
+
+            # select in order of priority:
+            # - a Primary config with project None (least prio)
+            # - a Primary config matching the project
+            # - a per user config with project None
+            # - a per user config matching the project (most prio)
+            pipeline_config = secondary_primary_config
+            if primary_config:
+                pipeline_config = primary_config
+            if secondary_user_config:
+                pipeline_config = secondary_user_config
+            if user_config:
+                pipeline_config = user_config
```

Comment:
This section is a little confusing. Wondering if it would help to interleave the comments into the conditionals? Also, should these be structured as `if/elif` statements?

```python
# select in order of priority:
if project_user_config:
    # - a per user config matching the project
    pipeline_config = project_user_config
elif project_primary_config:
    # - a Primary config matching the project
    pipeline_config = project_primary_config
elif site_user_config:
    # - a per user config with project None
    pipeline_config = site_user_config
elif site_primary_config:
    # - a Primary config with project None
    pipeline_config = site_primary_config
else:
    # - No pipeline config found, fallback to the base config
    pipeline_config = None
```

Maybe a little too explicit?